### ISSUE TITLE
Num_time_steps calculation for batch inputs is wrong

### ISSUE THREAD

**Issue by ankitmundada** [parlance/ctcdecode#54]
When using ctcdecode with sequential data of variable output lengths, the smaller outputs are generally padded with zeros to compensate for the extra size of the largest sample. So, logically, when the `ctc_beam_search_decoder` loops through the timesteps of `probs_seq` at [Link for code](https://github.com/parlance/ctcdecode/blob/master/ctcdecode/src/ctc_beam_search_decoder.cpp#L60), it should stop at the timestep corresponding to the actual size of that sample's output instead of the length of the `probs_seq`, since `probs_seq` also has extra padding in batch mode. This causes in ctcdecode to add extra _garbage_ characters at the end of its actual output. 

Examples of such outputs are:
```
Example#1:
Prediction: didn't do before a ooooooh i o o t h o e l l e e e e e e e e e e e e e e o a o o ghx xxx xxx eee e

Reference: didn't god before
```
```
Example#2:
Prediction: and it may be a lot of things that are kind of true ornette e e l e e e e e e e e e e e a u n ghx xxx eee et

Reference: and it may be a lot of things that are kind of truer
```
_I am using `ctcdecode` with the outputs from [deepspeech.pytorch](https://github.com/SeanNaren/deepspeech.pytorch/)_

I can think of two possible solutions for this:

1. Pass the `num_time_steps` to `ctc_beam_search_decoder` as an argument:
i.e. instead of
`size_t num_time_steps = probs_seq.size();`  at [line](https://github.com/parlance/ctcdecode/blob/master/ctcdecode/src/ctc_beam_search_decoder.cpp#L26), it should be 
`size_t num_time_steps = size # which is passed as an argument`

2. Add a check for some impossible probability outputs, such as `-1` and break the loop whenever its true.
_I am currently using this **hack** in our system, and it seems to work! You can find it [here](https://github.com/ankitmundada/ctcdecode/commit/5492a558107b4b3474b02d831997b2d6a41f1f34)_
_For this to work, the outputs of the DeepSpeech model are changed a bit. The extra timestep values are intentionally set to `-1`. The changes are [here](https://github.com/ankitmundada/deepspeech.pytorch/commit/e81aad01bd77aba6409ba2231cf8a73a06044ae8#diff-b284a28710cce90d9d9be3a7f4cabc8eR90)_

The transcripts for same examples, after using the second _hacky_ method are:
```
Example#1:
Prediction: didn't do before
Reference: didn't god before
```
```
Example#2:
Prediction: and it may be a lot of things that are kind of true or
Reference: and it may be a lot of things that are kind of truer
```

**Comment by ryanleary** [comment]
Thanks for the report and thorough investigation! I'll get a PR put together to address the issue.

**Comment by ryanleary** [comment]
@ankitmundada could you checkout the PR #55 and see if this rectifies your issue? You will have to change a line in `decoder.py` to pass the sequence lengths (see https://github.com/SeanNaren/deepspeech.pytorch/pull/239).

**Comment by ankitmundada** [comment]
@ryanleary I have tested it and it seems to work now! Thanks for the quick update!

**Comment by ryanleary** [comment]
Closed by #55.
